core/trino-main/src/main/java/io/trino/cost/TopNRankingStatsRule.java

```diff
@@ -0,0 +1,218 @@
+            // statistics. This makes the CBO estimates in the EXPLAIN plan output easier to understand,
+            // even though partial aggregations are added after the CBO rules have been run.            r
+            return Optional.of(sourceStats);
+        }
+
+        double sourceRowCount = sourceStats.getOutputRowCount();
+        // If sourceRowCount is 0, then the outputRowCount will also be 0.
+        if (sourceRowCount == 0) {
+            return Optional.of(PlanNodeStatsEstimate.buildFrom(sourceStats)
+                    .addSymbolStatistics(node.getRankingSymbol(), SymbolStatsEstimate.zero())
+                    .build());
+        }
+
+        double partitionCount = estimateCorrelatedDistinctCount(node.getPartitionBy(), sourceStats);
+        partitionCount = min(sourceRowCount, partitionCount);
+        if (isNaN(partitionCount)) {
+            // if partition count is NaN, we can't estimate anything, so we will return the source stats which is
+            // better than nothing.
+            return Optional.of(sourceStats);
+        }
+
+        // assuming no skew
+        double maxRowsPerPartition = sourceRowCount / partitionCount;
+
+        double rowsPerPartition;
+        double rankDistinctCount;
+        if (node.getRankingType() == ROW_NUMBER) {
+            // In case of ROW_NUMBER, each row gets a unique incremental rank. Therefore, the number of rows per
+            // partition can be at most the max rank per partition.
+            rowsPerPartition = min(maxRowsPerPartition, node.getMaxRankingPerPartition());
+            // The distinct count for ROW_NUMBER symbol should be equal to the number of rows per partition.
+            rankDistinctCount = rowsPerPartition;
+        }
+        else if (node.getRankingType() == RANK) {
+            // In case of RANK, each unique row gets a unique rank. However, unlike DENSE_RANK, the rank is not
+            // incremental.
+            // For example, ordering by ROW1, ROW2:
+            // ROW1   ROW2   RANK
+            // A      A      1
+            // A      A      1
+            // B      B      3
+            // B      B      3
+            // C      C      5
+            // C      C      5
+            // In the above example, the rank is not incremental. Therefore, the maximum number of rows per partition
+            // should be calculated using distinct count (assuming uniform distribution). In the above case
+            // rowsPerDistinctValue is 2 based on distinct count. Hence, the number of rows per partition for max rank
+            // 3 can be calculated using
+            // ceil(maxRank / rowsPerDistinctValue) * rowsPerDistinctValue = 4 => ceil(3/2) * 2 = 4 rows.
+            // Whereas, the distinct count for rank symbol should be 2 i.e. ceil(maxRank / rowsPerDistinctValue).
+            double distinctCount = estimateCorrelatedDistinctCount(node.getOrderingScheme().orderBy(), sourceStats);
```

Comment:
If you make it `min(estimateCorrelatedDistinctCount(node.getOrderingScheme().orderBy(), sourceStats), maxRowsPerPartition)` then if would feel more natural as we evaluate order by col distinct count *within* partition. `ceil` below also wouldn't be needed. We also avoid problem of dividing very small number with very large number

Please add a comment along the lines
`We assume order by columns are uncorrelated to partition columns, hence order by distinct rows should be equally spread among partitions`